docs/docs/typelevel.md

```diff
@@ -0,0 +1,449 @@
+# Functional Typelevel Programming in Scala
+
+This is a working draft document for discussing language constructs in
+support of typelevel programming in Scala 3.
+
+## State of the Art
+
+Currently, typelevel programming in Scala is mainly done using implicits.
+Proof search for finding implicit arguments can be used to compute new, interesting types.
+This results in a programming style much like Prolog. Amazing feats have
+been achieved using this scheme, but things are certainly far from
+ideal. In particular:
+
+ - The logic programming style requires a shift of mindset compared to the
+   usual functional programming style in Scala.
+ - The ways to control implicit search are underdeveloped,
+   leading to complicated schemes, requiring rule prioritization or other hacks.
+ - Because of their conceptual complexity the resulting typelevel programs are often
+   fragile.
+ - Error diagnostics are usually very bad, making programming with implicits somewhat
+   of a black art. Dotty has greatly improved error dignostics for recursive implicits,
+   but the fundamental problem remains.
+
+## The Core Idea
+
+A simple principle underlies the new design: Typelevel programming in Scala 3 means reducing terms and taking their types afterwards. Specifically, if `f` is a _transparent_ function applied to some arguments `es` then the type of the application `f(es)` is the type of the term to which `f(es)` reduces. Since reduction can specialize types, that type can be more specific than `f`'s declared result type. Type-level programming in Scala 3 is thus a form of partial evaluation or [type specialization](http://www.cse.chalmers.se/~rjmh/Papers/typed-pe.html).
+
+## Transparent Functions
+
+Consider the standard definition of typelevel Peano numbers:
+```scala
+trait Nat
+case object Z extends Nat
+case class S[N <: Nat](n: N) extends Nat
+```
+
+A function that maps positive integers to Peano numbers can be defined as follows:
+
+```scala
+transparent def toNat(n: Int): Nat = n match {
+  case 0 => Z
+  case n if n > 0 => S(toNat(n - 1))
+}
```

Comment:
It is absolutely a tradeoff but I want to register that to me at least, the ability to give an explicit return type that gets overridden has never felt right on whitebox macros and it still doesn't feel right to me here. Of course, I'm just one person and I can definitely get over it but I know that I, for one, would feel much more comfortable if you were required to leave off the return type on `transparent` methods and values.

Reply:
The return type is needed to do the initial type checking of the function (before inlining).

Also, if we consider `_` to have its usual semantics in types, which is that of an existential (and that's how I personally read it), then `_ <: Nat` or otherwise written `t forSome { type t <: Nat }` is exactly equivalent to `Nat` when used in positive positions (as here).

Reply:
Hrm.... It feels kind of weak for us to "need" an initial type for typechecking given the general theme of what `transparent` does. That said, I am not currently able to argue that it isn't needed, especially when we are dealing with recursion, which normally defeats method type inference.

I am taking us dangrously close to unproductive bikeshedding by suggesting this but what about `transparent def toNat(n: Int) <: Nat`? That should still parse unambiguously and to shows that something quite different from normal method type ascription is going on here?

Reply:
@clhodapp your version does read a little better, except for the flagrant conceptual violation that `<:` is for comparing types, and is really different from type ascription (in that sense, it would make Scala similar to these IMHO syntactically-misled languages which use a function-type-arrow to ascribe a result type to a method, as Rust with `fn foo(x: T) -> R { ... }`).
I think it's important that beginners understand that `<:` and `:` are fundamentally distinct concepts, and your proposed syntax would just muddy the waters for them.

Reply:
Fair. I throw my hands up on a syntax for now!

Reply:
Thought experiment -- what if it required a non-explicit (inferred) type? (Not suggesting as a requirement in actuality, for practical reasons, but I think it would be helpful if to do the thought experiment.)

Reply:
Here's another thought: require a type variable. Something like

`transparent def toNat[R <: Nat](n: Int): R = ???`

The "meaning" wouldn't have to change -- even in scala 2, what this means is, "compiler: please solve a type for R that has an upper bound of Nat," except that until now it could only use (unspecified) type inference rules (based on the constraints in the type signature). One way to look at this feature is, it's another case where you are giving the compiler a rough type but you want the compiler to drill it down more.

On the other hand, it's not actually a parameter. It isn't up to the caller to choose.

Hmm, almost like the difference between type parameters and type members. Except, methods can't have type members.

Oh well. Maybe food for thought anyway...